forecasting/src/autogluon/forecasting/trainer/abstract_trainer.py

```diff
@@ -0,0 +1,668 @@
+            model = self._train_single(
+                train_data, model, val_data=val_data, time_limit=time_limit
+            )
+            fit_end_time = time.time()
+
+            val_score = (
+                model.score(val_data) * METRIC_COEFFICIENTS[model.eval_metric]
+                if val_data is not None
+                else None
+            )
+            model.val_score = val_score
+
+            pred_end_time = time.time()
+
+            model.fit_time = model.fit_time or (fit_end_time - fit_start_time)
+            if model.predict_time is None:
+                model.predict_time = (
+                    None if val_score is None else (pred_end_time - fit_end_time)
+                )
+
+            self.save_model(model=model)
+
+        except Exception as err:
+            logger.error(
+                f"\tWarning: Exception caused {model.name} to fail during training... Skipping this model."
+            )
+            logger.error(f"\t\t{err}")
+            logger.exception("Detailed Traceback:")
+            del model
+        else:
+            self._add_model(model=model)  # noqa: F821
+            model_names_trained.append(model.name)  # noqa: F821
+            if self.low_memory:
+                del model
+
+        return model_names_trained
+
+    def _train_multi(
+        self,
+        train_data: TimeSeriesDataFrame,
+        hyperparameters: Optional[Union[str, Dict]] = None,
+        models: Optional[List[AbstractForecastingModel]] = None,
+        val_data: Optional[TimeSeriesDataFrame] = None,
+        hyperparameter_tune: bool = False,
+        time_limit: Optional[float] = None,
+    ) -> List[str]:
+        time_start = time.time()
+        if hyperparameters is not None:
+            self.hyperparameters = hyperparameters
+            hyperparameters = copy.deepcopy(hyperparameters)
+        else:
+            if models is None:
+                raise ValueError("Either models or hyperparameters should be provided")
+
+        if self.save_data and not self.is_data_saved:
+            self.save_train_data(train_data)
+            if val_data is not None:
+                self.save_val_data(val_data)
+            self.is_data_saved = True
+
+        if models is None:
+            models = self.construct_model_templates(
+                hyperparameters, hyperparameter_tune=hyperparameter_tune
+            )
+
+        time_limit_model_split = time_limit
+        if time_limit is not None and len(models) > 0:
+            time_limit_model_split /= len(models)
+
+        model_names_trained = []
+        for i, model in enumerate(models):
+            if hyperparameter_tune:
+                time_left = time_limit_model_split
+            else:
+                if time_limit is None:
+                    time_left = None
+                else:
+                    time_start_model = time.time()
+                    time_left = time_limit - (time_start_model - time_start)
```

Comment:
`if time_left <= 0` do we want to keep for looping or stop immediately with a log message? I believe Tabular stops immediately, but this code looks to continue looping.

Reply:
As far as I can tell Tabular keeps looping, but usually doesn't say it's "skipping due to lack of time remaining" because the log level is too low. I think it doesn't hurt to do this.

Reply:
I checked and Tabular stops immediately, it doesn't loop. I'd probably prefer this as it avoids redundant logging.

Reply:
That's strange. I'm running `example_simple_tabular.py` specifying 0.5 time limit to the predictor and also changing log level to 20 on https://github.com/awslabs/autogluon/blob/25f98756591c0cdb5344b575222e08bcef13e496/core/src/autogluon/core/trainer/abstract_trainer.py#L1054

The output: 
```
...
Fitting model: LightGBMXT ... Training model for up to 0.38s of the 0.38s of remaining time.
	Warning: Exception caused LightGBMXT to fail during training (ImportError)... Skipping this model.
		`import lightgbm` failed. A quick tip is to install via `pip install lightgbm`.
Fitting model: LightGBM ... Training model for up to 0.37s of the 0.37s of remaining time.
	Warning: Exception caused LightGBM to fail during training (ImportError)... Skipping this model.
		`import lightgbm` failed. A quick tip is to install via `pip install lightgbm`.
Fitting model: RandomForestGini ... Training model for up to 0.36s of the 0.36s of remaining time.
	0.81	 = Validation score   (accuracy)
	0.58s	 = Training   runtime
	0.04s	 = Validation runtime
Skipping RandomForestEntr due to lack of time remaining.
Skipping CatBoost due to lack of time remaining.
Skipping ExtraTreesGini due to lack of time remaining.
Skipping ExtraTreesEntr due to lack of time remaining.
Skipping NeuralNetFastAI due to lack of time remaining.
Skipping XGBoost due to lack of time remaining.
Skipping NeuralNetTorch due to lack of time remaining.
Skipping LightGBMLarge due to lack of time remaining.
```

Reply:
Anyway, I'll add a break here.